internal/webhooks/agent/pod_defaulter.go

```diff
@@ -0,0 +1,363 @@
+			},
+			corev1.EnvVar{
+				Name:  "CRYOSTAT_AGENT_WEBSERVER_TLS_CERT_TYPE",
+				Value: "X.509",
+			},
+			corev1.EnvVar{
+				Name:  "CRYOSTAT_AGENT_WEBSERVER_TLS_CERT_ALIAS",
+				Value: "cryostat",
+			},
+			corev1.EnvVar{
+				Name:  "CRYOSTAT_AGENT_WEBSERVER_TLS_KEY_PATH",
+				Value: fmt.Sprintf("/var/run/secrets/io.cryostat/cryostat-agent/%s", corev1.TLSPrivateKeyKey),
+			},
+			corev1.EnvVar{
+				Name:  "CRYOSTAT_AGENT_WEBSERVER_TLS_KEY_TYPE",
+				Value: "RSA",
+			},
+			corev1.EnvVar{
+				Name:  "CRYOSTAT_AGENT_WEBSERVER_TLS_KEY_ALIAS",
+				Value: "cryostat",
+			},
+		)
+	}
+	// Inject agent using JAVA_TOOL_OPTIONS, appending to any existing value
+	extended, err := extendJavaToolOptions(container.Env)
+	if err != nil {
+		return err
+	}
+	container.Env = extended
+
+	// Use GenerateName for logging if no explicit Name is given
+	podName := pod.Name
+	if len(podName) == 0 {
+		podName = pod.GenerateName
+	}
+	r.log.Info("configured Cryostat agent for pod", "name", podName, "namespace", pod.Namespace)
+
+	return nil
+}
+
+func cryostatURL(cr *model.CryostatInstance, tls bool) string {
+	// Build the URL to the agent proxy service
+	scheme := "https"
+	if !tls {
+		scheme = "http"
+	}
+	return fmt.Sprintf("%s://%s.%s.svc:%d", scheme, common.AgentProxyServiceName(cr), cr.InstallNamespace,
+		getAgentProxyHTTPPort(cr))
+}
+
+func getAgentProxyHTTPPort(cr *model.CryostatInstance) int32 {
+	port := constants.AgentProxyContainerPort
+	if cr.Spec.ServiceOptions != nil && cr.Spec.ServiceOptions.AgentConfig != nil && cr.Spec.ServiceOptions.AgentConfig.HTTPPort != nil {
+		port = *cr.Spec.ServiceOptions.AgentConfig.HTTPPort
+	}
+	return port
+}
+
+func (r *podMutator) callbackEnv(cr *model.CryostatInstance, namespace string, tls bool) []corev1.EnvVar {
+	scheme := "https"
+	if !tls {
+		scheme = "http"
+	}
+	envs := []corev1.EnvVar{
+		{
+			Name:  "CRYOSTAT_AGENT_CALLBACK_SCHEME",
+			Value: scheme,
+		},
+		{
+			Name:  "CRYOSTAT_AGENT_CALLBACK_HOST_NAME",
+			Value: fmt.Sprintf("$(%s), $(%s)[replace(\".\"\\, \"-\")]", podNameEnvVar, podIPEnvVar),
+		},
+		{
+			Name:  "CRYOSTAT_AGENT_CALLBACK_DOMAIN_NAME",
+			Value: fmt.Sprintf("%s.%s.svc", common.AgentHeadlessServiceName(r.gvk, cr), namespace),
+		},
+		{
+			Name:  "CRYOSTAT_AGENT_CALLBACK_PORT",
+			Value: "9977",
+		},
+	}
+
+	return envs
+}
+
+func (r *podMutator) getImageTag() string {
+	// Lazily look up image tag
+	if r.config.InitImageTag == nil {
+		agentInitImage := r.config.GetEnvOrDefault(agentInitImageTagEnv, constants.DefaultAgentInitImageTag)
+		r.config.InitImageTag = &agentInitImage
+	}
+	return *r.config.InitImageTag
+}
+
+func extendJavaToolOptions(envs []corev1.EnvVar) ([]corev1.EnvVar, error) {
+	existing, err := findJavaToolOptions(envs)
+	if err != nil {
+		return nil, err
+	}
+
+	if existing != nil {
+		existing.Value += " " + agentArg
+	} else {
+		envs = append(envs, corev1.EnvVar{
+			Name:  "JAVA_TOOL_OPTIONS",
```

Comment:
Maybe this could be another label-based extension point? Just in case the application is already setting this to some other value - ex in its Deployment, for the purpose of launching another Agent. The user might have another environment variable they can use for a similar purpose.

And also, maybe here we can check if the `JAVA_TOOL_OPTIONS` is already set in the container's env vars? If it is then I think at least emitting some warning would be good, but I think it's also likely to be OK to simply append our arg line to it rather than overwriting it.

Reply:
As I continue thinking on this I can't really come up with a scenario where this won't work. Maybe if the application has some strange entrypoint script and bootup logic that would cause this environment variable setting to get overridden on that side?